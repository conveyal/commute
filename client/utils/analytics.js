/* globals ga */

function init (i, s, o, g, r, a, m) {
  i['GoogleAnalyticsObject'] = r
  // eslint-disable-next-line
  i[r] = i[r] || function () {(i[r].q = i[r].q || []).push(arguments)}, i[r].l = 1 * new Date();
  // eslint-disable-next-line
  a = s.createElement(o),
    m = s.getElementsByTagName(o)[0]
  a.async = 1
  a.src = g
  m.parentNode.insertBefore(a, m)
}

function hasAnalytics () {
  if (process.env.NODE_ENV === 'development' || process.env.NODE_ENV === 'test') {
    return false
  } else if (!process.env.GOOGLE_ANALYTICS_TRACKING_ID) {
    console.error('GOOGLE_ANALYTICS_TRACKING_ID not set')
    return false
  } else {
    return true
  }
}

export function initialize () {
  if (!hasAnalytics()) return
  init(
    window,
    document,
    'script',
    'https://www.google-analytics.com/analytics.js',
    'ga'
  )

  ga('create', process.env.GOOGLE_ANALYTICS_TRACKING_ID, 'auto')
}

export function pageview (pageName) {
  if (!hasAnalytics()) return
  ga('send', 'pageview', pageName)
}
