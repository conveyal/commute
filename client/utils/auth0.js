import Auth0Lock from 'auth0-lock'

export function getLock (options={}) {
  return new Auth0Lock(
    process.env.AUTH0_CLIENT_ID,
    process.env.AUTH0_DOMAIN,
    options
  )
}
