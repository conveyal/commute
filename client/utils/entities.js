import update from 'react-addons-update'

/**
 * An entity is an object that has an id attribute
 * An entity map is a lookup of entities by id
 */

/**
 * Add a new entity to a entity map
 *
 * @param {Object} map       The entity map to add the entity to
 * @param {Object} newEntity The new entity
 * @return {Object}          The updated entity map
 */
export function addToEntityMap (map, newEntity) {
  return update(map, {
    [newEntity._id]: {
      $set: newEntity
    }
  })
}

/**
 * Add a many new entity to a entity map
 *
 * @param {Object} map          The entity map to add the entities to
 * @param {Object} newEntities  The new entities
 * @return {Object}             The updated entity map
 */
export function addEntitiesToEntityMap (map, newEntities) {
  let updatedMap = map
  newEntities.forEach((entity) => {
    updatedMap = addToEntityMap(updatedMap, entity)
  })
  return updatedMap
}

/**
 * Delete an entity from an entity map
 *
 * @param {Object} map      The entity map
 * @param {String} entityId The id of the entity to delete
 * @return {Object}         The updated entity map
 */
export function deleteFromEntityMap (map, entityId) {
  return update(map, {
    $apply: (obj) => {
      delete obj[entityId]
      return obj
    }
  })
}

/**
 * Delete an entity from an entity map
 *
 * @param {Object} map          The entity map
 * @param {String} queryParams  The query to match to delete
 * @return {Object}             The updated entity map
 */
export function deleteManyFromEntityMap (map, queryParams) {
  return update(map, {
    $apply: (entityMap) => entityArrayToEntityMap(
      Object.values(entityMap).filter((entity) => {
        for (const key in queryParams) {
          if (entity[key] !== queryParams[key]) {
            return true
          }
        }
        return false
      })
    )
  })
}

/**
 * Title says it all
 *
 * @param  {Array} arr The entity array
 * @return {Array}     An array of entity ids
 */
export function entityArrayToEntityIdArray (arr) {
  return arr.map((entity) => entity._id)
}

/**
 * Create an entity map from an array of entities
 *
 * @param  {Array} arr The array of entities
 * @return {Object}    An entity map
 */
export function entityArrayToEntityMap (arr) {
  const obj = {}
  for (let i = 0; i < arr.length; i++) obj[arr[i]._id] = arr[i]
  return obj
}

/**
 * Get the entities corresponding to a list of entity ids
 * entities not found in store will not be added to output array
 *
 * @param  {Array} entityIdArray An array of ids
 * @param  {Object} entityMap    The entity map to lookup the objects from
 * @return {Array}               An array of entities
 */
export function entityIdArrayToEntityArray (entityIdArray, entityMap) {
  const entities = []
  entityIdArray = entityIdArray || []
  for (let i = 0; i < entityIdArray.length; i++) {
    const curId = entityIdArray[i]
    if (entityMap[curId]) {
      entities.push(entityMap[curId])
    }
  }
  return entities
}
