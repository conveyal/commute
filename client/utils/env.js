export const messages = JSON.parse(process.env.MESSAGES)
export const settings = JSON.parse(process.env.SETTINGS)
