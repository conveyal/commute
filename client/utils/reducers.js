export function arrayToObj (a) {
  const obj = {}
  for (let i = 0; i < a.length; i++) obj[a[i].id] = a[i]
  return obj
}

/**
 * Round a number to a fixed amount of decimal places
 *
 * @param  {Number} n      The number to round
 * @param  {Number} places The number of places to round to (default=2)
 * @return {Number}        The rounded number
 */
export function fixedRound (n, places) {
  places = places || 2
  const multiplier = Math.pow(10, places)
  return Math.round(n * multiplier) / multiplier
}
