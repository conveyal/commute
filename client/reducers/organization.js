import update from 'react-addons-update'

import {settings} from '../utils/env'
import {arrayToObj, fixedRound} from '../utils/reducers'

const METRICS = Object.keys(settings.metrics)

export const reducers = {

  /**
   * Organization Stuff
   */

  'add organization' (state, action) {
    return addEntities(state, 'organizations', [action.payload])
  },
  'delete organization' (state, action) {
    return deleteEntity(state, 'organizations', action.payload.id)
  },
  'set organization' (state, action) {
    return setOrganization(state, action.payload)
  },
  'set organizations' (state, action) {
    const organizations = [...action.payload]
    return {
      organizations,
      organizationsById: arrayToObj(action.payload)
    }
  },

  /**
   * Analysis Stuff
   */

  'add analysis' (state, action) {
    const {organizationId, analysis} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = addEntities(affectedOrganization, 'analyses', [analysis])
    return setOrganization(state, updatedOrganization)
  },
  'delete analysis' (state, action) {
    const {organizationId, analysisId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = deleteEntity(affectedOrganization, 'analyses', analysisId)
    return setOrganization(state, updatedOrganization)
  },
  'receive mock calculated trips' (state, action) {
    const {analysisId, organizationId, trips} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const affectedAnalysis = affectedOrganization.analysesById[analysisId]
    let updatedAnalysis = update(affectedAnalysis, {
      trips: { $set: trips }
    })

    // calculate array of values for each metric
    const vals = {}
    const modes = Object.keys(trips[0]).filter((mode) => (['commuterId']).indexOf(mode) === -1)
    modes.forEach((mode) => {
      vals[mode] = {
        cost: [],
        distance: [],
        time: []
      }
    })
    trips.forEach((trip) =>
      modes.forEach((mode) =>
        METRICS.forEach((metric) => {
          vals[mode][metric].push(trip[mode][metric])
        })
      )
    )

    // sort arrays
    modes.forEach((mode) =>
      METRICS.forEach((metric) => {
        vals[mode][metric].sort()
      })
    )

    updatedAnalysis = update(updatedAnalysis, {
      tripVals: { $set: vals }
    })

    // calculate analysis summary metrics
    // Potential Savings is an overall average of the difference in cost between
    // the most likely trip and the driving trip
    const len = trips.length
    let totalCostDifference = 0
    let totalTravelTime = 0
    let totalDistance = 0
    trips.forEach((trip) => {
      totalCostDifference += trip.car.cost - trip.mostLikely.cost
      totalTravelTime += trip.mostLikely.time
      totalDistance += trip.mostLikely.distance
    })

    updatedAnalysis = update(updatedAnalysis, {
      summary: {
        $set: {
          avgTravelTime: fixedRound(totalTravelTime / len),
          avgDistance: fixedRound(totalDistance / len),
          savingsPerTrip: fixedRound(totalCostDifference / len),
          savingsPerTripYear: fixedRound(totalCostDifference / len * 260),
          savingsTotalPerDay: fixedRound(totalCostDifference),
          savingsTotalPerYear: fixedRound(totalCostDifference * 260)
        }
      }
    })

    const updatedOrganization = setEntity(affectedOrganization, 'analyses', updatedAnalysis)
    return setOrganization(state, updatedOrganization)
  },

  /**
   * Commuter Stuff
   */

  'add commuter' (state, action) {
    const {commuter, groupId, organizationId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const affectedGroup = affectedOrganization.groupsById[groupId]
    const updatedGroup = addEntities(affectedGroup, 'commuters', [commuter])
    const updatedOrganization = setGroup(affectedOrganization, updatedGroup)
    return setOrganization(state, updatedOrganization)
  },
  'delete commuter' (state, action) {
    const {commuterId, groupId, organizationId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const affectedGroup = affectedOrganization.groupsById[groupId]
    const updatedGroup = deleteEntity(affectedGroup, 'commuters', commuterId)
    const updatedOrganization = setGroup(affectedOrganization, updatedGroup)
    return setOrganization(state, updatedOrganization)
  },
  'update commuter' (state, action) {
    const {commuter, groupId, organizationId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const affectedGroup = affectedOrganization.groupsById[groupId]
    const updatedGroup = setEntity(affectedGroup, 'commuters', commuter)
    const updatedOrganization = setGroup(affectedOrganization, updatedGroup)
    return setOrganization(state, updatedOrganization)
  },

  /**
   * Group Stuff
   */

  'add group' (state, action) {
    const {organizationId, group} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    group.commutersById = arrayToObj(group.commuters)
    const updatedOrganization = addEntities(affectedOrganization, 'groups', [group])
    return setOrganization(state, updatedOrganization)
  },
  'append commuters' (state, action) {
    const {commuters, groupId, organizationId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const affectedGroup = affectedOrganization.groupsById[groupId]
    const updatedGroup = addEntities(affectedGroup, 'commuters', commuters)
    const updatedOrganization = setGroup(affectedOrganization, updatedGroup)
    return setOrganization(state, updatedOrganization)
  },
  'delete group' (state, action) {
    const {organizationId, groupId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = deleteEntity(affectedOrganization, 'groups', groupId)
    return setOrganization(state, updatedOrganization)
  },

  /**
   * Site Stuff
   */

  'add site' (state, action) {
    const {organizationId, site} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = addEntities(affectedOrganization, 'sites', [site])
    return setOrganization(state, updatedOrganization)
  },
  'delete site' (state, action) {
    const {organizationId, siteId} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = deleteEntity(affectedOrganization, 'sites', siteId)
    return setOrganization(state, updatedOrganization)
  },
  'update site' (state, action) {
    const {organizationId, site} = action.payload
    const affectedOrganization = state.organizationsById[organizationId]
    const updatedOrganization = setEntity(affectedOrganization, 'sites', site)
    return setOrganization(state, updatedOrganization)
  }
}

// Utilities

/**
 * Helper to delete an item from a entityById object
 * @param  {String} id  The ID of the entity to delete
 * @return {Function}   A function that deletes the entity from
 *  the object and then returns that object
 */
const deleteFromMap = (id) => (map) => {
  delete map[id]
  return map
}

export const initialState = {
  organizations: [],
  organizationsById: {}
}

/**
 * Add entities to a collection
 *
 * @param {Object} parent           the parent object containing the collections
 * @param {String} entitiesName     base name for entity collection keys
 * @param {Array} entities          the entities to add
 * @return                          The updated parent
 */
const addEntities = (parent, entitiesName, entities) => {
  const entitiesByIdName = `${entitiesName}ById`
  let updatedParent = update(parent, {
    [entitiesName]: { $push: entities }
  })
  return update(updatedParent, {
    [entitiesByIdName]: { $set: arrayToObj(updatedParent[entitiesName]) }
  })
}

/**
 * Delete an entity within a collection
 *
 * @param {Object} parent           the parent object containing the collections
 * @param {String} entitiesName     base name for entity collection keys
 * @param {Object} entityId         ID of the entity to delete
 * @return                          The updated parent
 */
const deleteEntity = (parent, entitiesName, entityId) => {
  const entitiesByIdName = `${entitiesName}ById`
  let updatedParent = update(parent, {
    [entitiesByIdName]: { $apply: deleteFromMap(entityId) }
  })
  return update(updatedParent, {
    [entitiesName]: { $set: Object.values(updatedParent[entitiesByIdName]) }
  })
}

/**
 * Helper to update an entity within a collection
 *
 * @param {Object} parent           the parent object containing the collections
 * @param {String} entitiesName     base name for entity collection keys* @param {Object} updatedEntity    entity to update
 * @return                          The updated parent
 */
const setEntity = (parent, entitiesName, updatedEntity) => {
  const entitiesByIdName = `${entitiesName}ById`
  let updatedParent = update(parent, {
    [entitiesByIdName]: {
      [updatedEntity.id]: {
        $set: updatedEntity
      }
    }
  })
  return update(updatedParent, {
    [entitiesName]: {
      $set: Object.values(updatedParent[entitiesByIdName])
    }
  })
}

const setGroup = (organization, group) => {
  return setEntity(organization, 'groups', group)
}

const setOrganization = (state, organization) => {
  return setEntity(state, 'organizations', organization)
}
