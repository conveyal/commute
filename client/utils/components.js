import {settings} from './env'

/**
 * Calculate number of elements in array less than or equal to target values
 *
 * @param  {Number[]} arr     A sorted array of numbers
 * @param  {[Number]} target  The number to be less than or equal to
 * @return {int}              The number of elements in the array less than or equal to the target value
 */
export function calcNumLessThan (arr, target) {
  // do binary search
  let left = 0
  let right = arr.length - 1
  let mid

  while (left <= right) {
    mid = Math.floor((left + right) / 2)
    if (arr[mid] <= target) {
      left = mid + 1
    } else {
      right = mid - 1
    }
  }

  return left
}

/**
 * Helper function to get initial series data for drawing graphs
 *
 * @return {Obj[]} Array of objects with series template
 */
export function getInitialSeries () {
  const modes = Object.keys(settings.modeDisplay)
  const series = []

  modes.forEach((mode) => {
    series.push(Object.assign({disabled: false, mode}, settings.modeDisplay[mode]))
  })

  return series
}

/**
 * Humanize distances
 * // TODO: more humanization
 *
 * @param  {float} d Distance to humanize
 * @return {string} humanized distance
 */
export function humanizeDistance (d) {
  return `${d} miles`
}
