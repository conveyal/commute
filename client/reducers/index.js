import {handleActions} from 'redux-actions'

import * as organization from './organization'
import * as user from './user'

export default {
  organization: handleActions(organization.reducers, organization.initialState),
  user: handleActions(user.reducers, user.initialState)
}
