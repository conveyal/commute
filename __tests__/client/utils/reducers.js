/* globals describe, expect, it */

import * as reducers from '../../../client/utils/reducers'

describe('utils > reducers', () => {
  it('arrayToObj should work', () => {
    expect(reducers.arrayToObj([{id: '1'}])).toMatchSnapshot()
  })
})
