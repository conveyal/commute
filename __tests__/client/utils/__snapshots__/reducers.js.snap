exports[`utils > reducers arrayToObj should work 1`] = `
Object {
  "1": Object {
    "id": "1",
  },
}
`;
